File: java/ch17/src/test/java/com/thesoftwaregorilla/tdd/money/CurrencyPairTest.java
package com.thesoftwaregorilla.tdd.money;

import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("CurrencyPair Tests")
public class CurrencyPairTest {

    @Test
    @DisplayName("equal")
    public void testCurrencyPairEquals() {
        var francToDollar = new CurrencyPair("CHF", "USD");
        assertEquals(francToDollar, new CurrencyPair("CHF", "USD"));
        assertEquals(francToDollar.hashCode(), new CurrencyPair("CHF", "USD").hashCode());
    }

    @Test
    @DisplayName("not equal")
    public void testCurrencyPairNotEquals() {
        var francToDollar = new CurrencyPair("CHF", "USD");
        assertNotEquals(francToDollar, new CurrencyPair("USD", "CHF"));
        assertNotEquals(francToDollar.hashCode(), new CurrencyPair("USD", "CHF").hashCode());
    }

    @Test
    @DisplayName("edge tests for code coverage")
    public void testCurrencyEdgeCases() {
        var francToDollar = new CurrencyPair("CHF", "USD");
        // Test for code coverage
        assertEquals(francToDollar, francToDollar);
        assertNotEquals(francToDollar, "CHF");
    }

}

File: cs/ch17-01-tests/Money.Tests/CurrencyPairTest.cs
using Microsoft.VisualStudio.TestPlatform.ObjectModel;
using NUnit.Framework;
using TheSoftwareGorilla.TDD.Money;

namespace TheSoftwareGorilla.TDD.Money.Tests;

#region TO DO's
#endregion

[TestFixture]
[Description("CurrencyPair Tests")]
public class CurrencyPairTest
{

    [TestCase("CHF", "USD", TestName = "CurrencyPair CHF to USD are equal")]
    [TestCase("ZAR", "USD", TestName = "CurrencyPair ZAR to USD are equal")]
    [TestCase("CHF", "USD", false, TestName = "CurrencyPair CHF to USD != USD to CHF")]
    [TestCase("ZAR", "USD", false, TestName = "CurrencyPair ZAR to USD != USD to ZAR")]
    public void TestCurrencyPairs(string from, string to, bool isEqual = true)
    {
        var pair = new CurrencyPair(from, to);
        if(isEqual)
        {
            Assert.That(pair, Is.EqualTo(new CurrencyPair(from, to)));
            Assert.That(pair.GetHashCode(), Is.EqualTo(new CurrencyPair(from, to).GetHashCode()));
        }
        else
        {
            Assert.That(pair, Is.Not.EqualTo(new CurrencyPair(to, from))); // Constructor parameter order intentional to test inequality
            Assert.That(pair.GetHashCode(), Is.Not.EqualTo(new CurrencyPair(to, from).GetHashCode())); // Constructor parameter order intentional to test inequality

        }
    }   

    [TestCase("CHF", "USD", TestName = "CurrencyPair CHF to USD edge case")]
    [TestCase("ZAR", "USD", TestName = "CurrencyPair ZAR to USD edge case")]
    [Description("Edge cases for code coverage")]
    public void TestEdgeCases(string from, string to)
    {
        var pair = new CurrencyPair(from, to);
        Assert.That(pair, Is.Not.EqualTo(null));
        Assert.That(pair, Is.Not.EqualTo(new object()));
        Assert.That(pair, Is.EqualTo(pair));
    }
}